amads/polyphony/skyline.py

```diff
@@ -41,51 +41,89 @@ def skyline(score: Score, threshold: float = 0.1):
-    score = score.strip_ties()
-    if not score.is_flattened_and_collapsed():
-        score = score.flatten(collapse=True)
-    filtered_notes = []
-    notes = list(score.find_all(Note))
-
-    # sort the notes by start, if start is equal, sort by pitch
-    notes.sort(key=lambda note: (note.onset, -note.pitch.keynum))
+    score = score.merge_tied_notes()
+    new_score = score.emptycopy()
+    skyline = Part(new_score)
+    notes = score.sorted_notes()
+
+    # Algorithm: the basic idea is to scan notes and copy them
+    # to skyline, a Part object. We can use shallow copy
+    # because notes are already deep copied from score after
+    # merge_tied_notes.
+    #
+    # In the outer looop, we test each note to see if it is below
+    # the skyline as it exists so far. Since we process in order,
+    # we know each note starts after all notes in the skyline, so
+    # time overlap occurs when a note starts before the end of a
+    # skyline note. (Use a threshold in case the note starts at
+    # approximately the previous note end time) When a note is
+    # added to the skyline, it may be above an existing skyline
+    # note that started earlier, so the inner loop deals with
+    # this problem.
+    #
+    # In the inner loop, we remove skyline notes that started before
+    # the new note and are below it (if they were above, the new note
+    # would have been filtered out before we got to this part). Since
+    # notes have arbitrary duration, *any* skuyline note could potentially
+    # overlap in time and be below the new note, so we have to check
+    # all skyline notes. We search in reverse order so that we can remove
+    # notes from skyline without worrying about the index changing.
+    # Since we know the new note starts after all skyline notes, again we
+    # can test for time overlap by checking if the new note starts before
+    # the end of the skyline note. We remove the lower note only when
+    # the new note starts within threshold of the skuyline note. Otherwise,
+    # the lower note plays for longer than threshold before the new note:
+    # we shorten the duration of the lower note to end at the onset time
+    # of the new note.
+    #
+    # A consequence of this algorithm is that a very long low note will
+    # be shortened to the onset time of a new note, so a piano roll like
+    # this:                      ----------
+    #          ------------------------------------------
+    # will result in this:       ----------
+    #          ------------------          (nothing here)
+    # rather than this:          ----------
+    #          ------------------          --------------
+    # A faster algorithm would avoid iterating over all skyline notes
+    # (twice!) since we know the skyline is monophonic and ordered.
 
     for i in range(len(notes)):
         note = notes[i]
-        # ignore notes that are below another existing note in filtered_notes
+        print("Processing note ", end="")
```

Comment:
It might be good for us to start using `logger.debug` for these kinds of print statements

Reply:
but I guess actually you were planning on deleting these

Reply:
Mostly, these are debugging statements I forgot to remove. I should search for "print". I think skyline needs to be compared to MIDI Toolbox -- and it's not obvious how all the edge cases should be handled.